source/documentation/adrs/adr-009.html.md.erb

```diff
@@ -0,0 +1,70 @@
+---
+title: ADR-009 Log structure
+last_reviewed_on: 2024-02-09
+review_in: 12 months
+---
+# <%= current_page.data.title %>
+
+Date: 14-02-2024
+
+## Status
+
+🤔 Proposed
+
+## Context
+
+Regardless of platform or language, our services write logging information to the output stream. However, this logging information is inconsitently structured. This leads to a few problems:
+
+1. It is hard to interpret logs under standard operation, as logs from different containers with different structures are interspersed
+2. During an incident, it is hard to find related logs as the inconsistent structures require separate queries
+3. Our log-based alarms are complex to write and update as they have to query several different structures
+
+## Decision
+
+We will define a standard logging structure that all teams will use. Teams are welcome to expand on the structure by adding extra fields, but must include all required fields in each log message and must ensure not to use reserved fields for anything other than their defined usage.
+
+All logs must be formatted as JSON objects.
+
+An example of a minimal log message:
+
+```json
+{"timestamp": "2024-02-14T12:34:23Z", "level": "CRITICAL", "msg": "Null pointer exception"}
+```
+
+An example of a more complex log message:
+
+```json
+{
+  "timestamp": "2024-02-14T13:39:01Z",
+  "level": "INFO",
+  "msg": "User permissions updated",
+  "trace_id": "1-581cf771-a006649127e371903a2de979",
+  "request": { "method": "PUT", "path": "/user/133/permissions" },
+  "location": { "file": "pages/user/edit_permission.go", "line": 156 },
+  "actor_id": 48
+}
+```
+
+As well as the required and reserved fields below, log messages can include any other fields (such as "location" and "actor_id" above).
+
+### Required fields
+
+`timestamp` must be an RFC3339-formatted date/time.
+
+`level` must be an RFC5424-defined log level in uppercase: `EMERGENCY`, `ALERT`, `CRITICAL`, `ERROR`, `WARNING`, `NOTICE`, `INFO`, `DEBUG`. Services don't need to use every log level, but alarms and stored queries should account for them all.
+
+`msg` is the free form message.
+
+### Reserved fields
+
+`request`, if provided, must be an object containing a `method` property of the request's HTTP method, and a `path` property of the requested path. Additional properties are permitted.
+
+`source`, if provided, must be a string that differntiates which resource the log came from. It should be set consistently at the resource level (e.g. all logs from one ECS container should have the same source).
```

Comment:
differntiates -> differentiates

Reply:
(not trying to be like, let `log/slog` dictate things) but it uses `source` as the key for where in the code a line was logged,  e.g.

```
...,"source":{"function":"main.run","file":"/app/cmd/mlpa/main.go","line":353},...
```

Could we log this as `resource` (or maybe `arn`) instead?

Reply:
That's fair enough, we definitely don't want a policy that actively conflicts with default settings of existing tools.

But I've removed this entirely, because it's poorly defined and not necessary with `@log` and `@logStream` available. Teams may want to do something similar, but if we can't define it clearly at an org-level then it doesn't seem worth mentioning here.